README.md
### Legacy Tracker
Legacy Tracker is a Nodejs Fork of Minetrack that has been developed to be compatible with JSON APIS.
This has proven extremely useful with Legacy versions of the game that do now have the Modern day Ping System.


Live Version: https://tracker.johnymuffin.com

CHANGELOG
Update Atlis-0.4
    Restore Ping Logic
        Add Dynmap Ping Type
        Add Modern Ping Type Back
Update Atlis-0.3
    Average Players Instead Of Top
    Use Cloudflare For IP
    Readd Footer
Update Atlis-0.2
    Enforce Favicon Size
Update Atlis-0.1
    Support For Beta

app.js
var server = require('./lib/server');
var ping = require('./lib/ping');
var logger = require('./lib/logger');
var mojang = require('./lib/mojang_services');
var util = require('./lib/util');
var db = require('./lib/database');

var config = require('./config.json');
var servers = require('./servers.json');

var networkHistory = [];
var connectedClients = 0;

var currentVersionIndex = {
	'PC': 0,
	'PE': 0
};

var networkVersions = [];

var graphData = [];
var highestPlayerCount = {};
var averages = {};
var lastGraphPush = [];

function pingAll() {
	for (var i = 0; i < servers.length; i++) {
		// Make sure we lock our scope.
		(function (network) {
			// Asign auto generated color if not present
			if (!network.color) {
				network.color = util.stringToColor(network.name);
			}

			var attemptedVersion = config.versions[network.type][currentVersionIndex[network.type]];
			ping.ping(network.ip, network.port, network.type, config.rates.connectTimeout, function (err, res) {
				// Handle our ping results, if it succeeded.
				if (err) {
					logger.log('error', 'Failed to ping ' + network.ip + ': ' + err.message);
				}

				// If we have favicon override specified, use it.
				if (res && config.faviconOverride && config.faviconOverride[network.name]) {
					res.favicon = config.faviconOverride[network.name];
				}

				handlePing(network, res, err, attemptedVersion);
			}, attemptedVersion);
		})(servers[i]);
	}

	currentVersionIndex['PC']++;
	currentVersionIndex['PE']++;

	if (currentVersionIndex['PC'] >= config.versions['PC'].length) {
		// Loop around
		currentVersionIndex['PC'] = 0;
	}

	if (currentVersionIndex['PE'] >= config.versions['PE'].length) {
		// Loop around
		currentVersionIndex['PE'] = 0;
	}
}

// This is where the result of a ping is feed.
// This stores it and converts it to ship to the frontend.
function handlePing(network, res, err, attemptedVersion) {
	// Log our response.
	if (!networkHistory[network.name]) {
		networkHistory[network.name] = [];
	}

	// Update the version list
	if (!networkVersions[network.name]) {
		networkVersions[network.name] = [];
	}

	// If the result version matches the attempted version, the version is supported
	var _networkVersions = networkVersions[network.name];
	if (res) {
		if (res.version == attemptedVersion) {
			if (_networkVersions.indexOf(res.version) == -1) {
				_networkVersions.push(res.version);
			}
		} else {
			// Mismatch, so remove the version from the supported version list
			var index = _networkVersions.indexOf(attemptedVersion);
			if (index != -1) {
				_networkVersions.splice(index, 1);
			}
		}
	}

	// Update the clients
	var networkSnapshot = {
		info: {
			name: network.name,
			timestamp: util.getCurrentTimeMs(),
			type: network.type
		},
		versions: _networkVersions,
		average: averages[network.ip],
		record: highestPlayerCount[network.ip]
	};

	if (res) {
		networkSnapshot.result = res;

		// Validate that we have logToDatabase enabled otherwise in memory pings
		// will create a record that's only valid for the runtime duration.

		if (config.logToDatabase && res.players.online > highestPlayerCount[network.ip]) {
			highestPlayerCount[network.ip] = res.players.online;
		}

		//Not gonna lie, i don't understand how to use SqLite but its not crashing so :)

		//Averages 
		db.getAverage(network.ip, function (record) {
			logger.log('info', 'Completed query for %s', network.ip);

			averages[network.ip] = record;
			server.io.sockets.emit('update', networkSnapshot);
		});

	} else if (err) {
		networkSnapshot.error = err;
	}

	//  server.io.sockets.emit('update', networkSnapshot);

	var _networkHistory = networkHistory[network.name];

	// Remove our previous data that we don't need anymore.
	for (var i = 0; i < _networkHistory.length; i++) {
		delete _networkHistory[i].info;

		if (_networkHistory[i].result) {
			delete _networkHistory[i].result.favicon;
		}
	}

	_networkHistory.push({
		error: err,
		result: res,
		versions: _networkVersions,
		timestamp: util.getCurrentTimeMs(),
		info: {
			ip: network.ip,
			port: network.port,
			type: network.type,
			name: network.name
		}
	});

	// Make sure we never log too much.
	if (_networkHistory.length > 72) { // 60/2.5 = 24, so 24 is one minute
		_networkHistory.shift();
	}

	// Log it to the database if needed.
	if (config.logToDatabase) {
		db.log(network.ip, util.getCurrentTimeMs(), res ? res.players.online : 0);
	}

	// Push it to our graphs.
	var timeMs = util.getCurrentTimeMs();

	// The same mechanic from trimUselessPings is seen here.
	// If we dropped the ping, then to avoid destroying the graph, ignore it.
	// However if it's been too long since the last successful ping, we'll send it anyways.
	if (config.logToDatabase) {
		if (!lastGraphPush[network.ip] || (timeMs - lastGraphPush[network.ip] >= 60 * 1000 && res) || timeMs - lastGraphPush[network.ip] >= 70 * 1000) {
			lastGraphPush[network.ip] = timeMs;

			// Don't have too much data!
			util.trimOldPings(graphData);

			if (!graphData[network.name]) {
				graphData[network.name] = [];
			}

			graphData[network.name].push([timeMs, res ? res.players.online : 0]);

			// Send the update.
			server.io.sockets.emit('updateHistoryGraph', {
				ip: network.ip,
				name: network.name,
				players: (res ? res.players.online : 0),
				timestamp: timeMs
			});
		}
	}
}

// Start our main loop that does everything.
function startMainLoop() {
	util.setIntervalNoDelay(pingAll, config.rates.pingAll);

	util.setIntervalNoDelay(function () {
		mojang.update(config.rates.mojangStatusTimeout);

		server.io.sockets.emit('updateMojangServices', mojang.toMessage());
	}, config.rates.upateMojangStatus);
}

function startServices() {
	server.start();

	// Track how many people are currently connected.
	server.io.on('connect', function (client) {
		// We're good to connect them!
		connectedClients += 1;

		logger.log('info', '%s connected, total clients: %d', client.request.connection.remoteAddress, connectedClients);

		// We send the boot time (also sent in publicConfig.json) to the frontend to validate they have the same config.
		// If so, they'll send back "requestListing" event, otherwise they will pull the new config and retry.
		client.emit('bootTime', util.getBootTime());

		// Attach our listeners.
		client.on('disconnect', function () {
			connectedClients -= 1;

			logger.log('info', '%s disconnected, total clients: %d', client.request.connection.remoteAddress, connectedClients);
		});

		client.on('requestHistoryGraph', function () {
			if (config.logToDatabase) {
				// Send them the big 24h graph.
				client.emit('historyGraph', graphData);
			}
		});

		client.on('requestListing', function () {
			// Send them our previous data, so they have somewhere to start.
			client.emit('updateMojangServices', mojang.toMessage());

			// Remap our associative array into just an array.
			var networkHistoryKeys = Object.keys(networkHistory);

			networkHistoryKeys.sort();

			// Send each individually, this should look cleaner than waiting for one big array to transfer.
			for (var i = 0; i < servers.length; i++) {
				var server = servers[i];

				if (!(server.name in networkHistory) || networkHistory[server.name].length < 1) {
					// This server hasn't been ping'd yet. Send a hacky placeholder.
					client.emit('add', [
						[{
							error: {
								description: 'Waiting'
							},
							result: null,
							timestamp: util.getCurrentTimeMs(),
							info: {
								ip: server.ip,
								port: server.port,
								type: server.type,
								name: server.name
							}
						}]
					]);
				} else {
					client.emit('add', [networkHistory[networkHistoryKeys[i]]]);
				}
			}

			client.emit('syncComplete');
		});
	});

	startMainLoop();
}

logger.log('info', 'Booting, please wait...');

if (config.logToDatabase) {
	// Setup our database.
	db.setup();

	var timestamp = util.getCurrentTimeMs();

	db.queryPings(config.graphDuration, function (data) {
		graphData = util.convertServerHistory(data);
		completedQueries = 0;

		logger.log('info', 'Queried and parsed ping history in %sms', util.getCurrentTimeMs() - timestamp);

		for (var i = 0; i < servers.length; i++) {
			(function (server) {
				db.getTotalRecord(server.ip, function (record) {
					logger.log('info', 'Completed query for %s', server.ip);

					highestPlayerCount[server.ip] = record;

					completedQueries += 1;

					if (completedQueries === servers.length) {
						//completedQueries = 0;

						//Get Averages
						db.getAverage(server.ip, function (record) {
							logger.log('info', 'Completed query for %s', server.ip);
		
							averages[server.ip] = record;
		
							//completedQueries += 1;
		
							//if (completedQueries === servers.length) {
								startServices();
							//}
						});

					}
				});
				

			})(servers[i]);
		}
	});
} else {
	logger.log('warn', 'Database logging is not enabled. You can enable it by setting "logToDatabase" to true in config.json. This requires sqlite3 to be installed.');

	startServices();
}

lib/ping.js
var mcpe_ping = require('mcpe-ping-fixed');
var mcpc_ping = require('mc-ping-updated');
var getJSON = require('get-json')

var util = require('./util');

// This is a wrapper function for mc-ping-updated, mainly used to convert the data structure of the result.
// function pingMinecraftPC(host, port, timeout, callback, version) {
// 	var startTime = util.getCurrentTimeMs();

// 	mcpc_ping(host, port, function (err, res) {
// 		if (err) {
// 			callback(err, null);
// 		} else {
// 			// Remap our JSON into our custom structure.
// 			callback(null, {
// 				players: {
// 					online: res.players.online,
// 					max: res.players.max
// 				},
// 				version: res.version.protocol,
// 				latency: util.getCurrentTimeMs() - startTime,
// 				favicon: res.favicon
// 			});
// 		}
// 	}, timeout, version);
// }

// This is a wrapper function for mcpe-ping, mainly used to convert the data structure of the result.
function pingMinecraftPE(host, port, timeout, callback) {
	var startTime = util.getCurrentTimeMs();

	mcpe_ping(host, port || 19132, function (err, res) {
		if (err) {
			callback(err, null);
		} else {
			// Remap our JSON into our custom structure.
			console.log(res.version)
			callback(err, {
				players: {
					online: parseInt(res.currentPlayers),
					max: parseInt(res.maxPlayers)
				},
				version: res.version,
				latency: util.getCurrentTimeMs() - startTime
			});
		}
	}, timeout);
}

function pingMinecraftPC(host, port, timeout, callback, version) {
    var startTime = util.getCurrentTimeMs();

	mcpc_ping(host, port, function(err, res) {
	    if (err) {
	        callback(err, null);
	    } else {
	    	// Remap our JSON into our custom structure.
	        callback(null, {
				players: {
					online: res.players.online,
					max: res.players.max
				},
				version: res.version.protocol,
				latency: util.getCurrentTimeMs() - startTime,
				favicon: res.favicon
			});
	    }
	}, timeout, version);
}

// Ping a Json Server for onlineCount
function pingMinecraftJson(host, port, timeout, callback) {
	var startTime = util.getCurrentTimeMs();

	getJSON(host, function (error, res) {
		if (error) {
			callback(error, null);
		} else {
			callback(error, {
				players: {
					online: parseInt(res.onlineCount),
					max: parseInt("100")
				},
				version: 0,
				latency: parseInt(100)
			});
		}
	})
}
//Ping Dynmap
function pingDynmap(host, port, timeout, callback) {
	var startTime = util.getCurrentTimeMs();

	getJSON(host, function (error, res) {
		if (error) {
			callback(error, null);
		} else {
			//Is Dynmap New
			let time = parseInt(res.timestamp) - Date.now();
			time = time/1000
			time = Math.abs(time)

			if(time < 120){
				callback(error, {
					players: {
						online: parseInt(res.players.length),
						max: parseInt("100")
					},
					version: 0,
					latency: parseInt(100)
				});
			} else {
				callback("Old Dynmap Cache", null);
			}

		}
	})
}


exports.ping = function (host, port, type, timeout, callback, version) {
	if (type === 'PC') {
		util.unfurlSRV(host, port, function (host, port) {
			pingMinecraftPC(host, port || 25565, timeout, callback, version);
		})
	} else if (type === 'PE') {
		pingMinecraftPE(host, port || 19132, timeout, callback);
	} else if (type === 'JSON') {
		pingMinecraftJson(host, port || 19132, timeout, callback);
	} else if (type === 'DYNMAP') {
		pingDynmap(host, port || 19132, timeout, callback);
	} else {
		throw new Error('Unsupported type: ' + type);
	}
};

lib/server.js
var http = require('http');
var fs = require('fs');
var url = require('url');
var mime = require('mime');
var io = require('socket.io');

var util = require('./util');
var logger = require('./logger');

var config = require('../config.json');
var minecraft = require('../minecraft.json');
var servers = require('../servers.json');

var urlMapping = [];

function setupRoutes() {
	var routeKeys = Object.keys(config.routes);

	// Map the (static) routes from our config.
	for (var i = 0; i < routeKeys.length; i++) {
		urlMapping[routeKeys[i]] = config.routes[routeKeys[i]];
	}

	logger.log('info', 'Routes: %s', routeKeys);
}

function handleRequest(req, res) {
	var requestUrl = url.parse(req.url).pathname;
	let ip = req.headers['cf-connecting-ip'] || req.headers['x-forwarded-for'] || req.connection.remoteAddress
	logger.log('info', '%s requested: %s', ip, requestUrl);

	if (requestUrl === '/status.json') {
		res.setHeader('Content-Type', 'text/plain');

		res.write(JSON.stringify({
			error: true,
			message: 'API deprecated.'
		}));

		res.end();
	} else if (requestUrl === '/publicConfig.json') {
		res.setHeader('Content-Type', 'application/javascript');

		var categories = config.serverCategories;

		// Legacy support for people without categories configured.
		if (!categories || Object.keys(categories).length === 0) {
			categories = {
				'default': 'All Networks'
			};
		}

		for (var i = 0; i < servers.length; i++) {
			var entry = servers[i];

			if (!entry.category) {
				entry.category = 'default';

				logger.warn('%s has no category, defaulting!', entry.name);
			} else if (!categories[entry.category]) {
				logger.warn('%s has an unknown category (%s), defaulting!', entry.name, entry.category);

				entry.category = 'default';
			}
		}

		var publicConfig = {
			categories: categories,
			graphDuration: config.graphDuration,
			servers: servers,
			bootTime: util.getBootTime(),
			categoriesVisible: config.categoriesVisible || false,
			serverTypesVisible: config.serverTypesVisible || false,
			minecraftVersions: minecraft.versions
		};

		res.write('setPublicConfig(' + JSON.stringify(publicConfig) + ');');

		res.end();
	} else if (requestUrl in urlMapping) {
		var file = urlMapping[requestUrl];

		res.setHeader('Content-Type', mime.lookup(file));

		fs.createReadStream(file).pipe(res);
	} else {
		res.statusCode = 404;

		res.write('404');

		res.end();
	}
}

exports.start = function() {
	setupRoutes();

	// Create our tiny little HTTP server.
	var server = http.createServer(handleRequest);

	server.listen(config.site.port, config.site.ip);

	// I don't like this. But it works, I think.
	exports.io = io.listen(server);

	// Since everything is loaded, let's celebrate!
	logger.log('info', 'Started on %s:%d', config.site.ip, config.site.port);
};


lib/database.js
var util = require('./util');
var config = require('../config.json');

exports.setup = function () {
	var sqlite = require('sqlite3');

	var db = new sqlite.Database('database.sql');

	db.serialize(function () {
		db.run('CREATE TABLE IF NOT EXISTS pings (timestamp BIGINT NOT NULL, ip TINYTEXT, playerCount MEDIUMINT)');
		db.run('CREATE INDEX IF NOT EXISTS ip_index ON pings (ip, playerCount)');
		db.run('CREATE INDEX IF NOT EXISTS timestamp_index on PINGS (timestamp)');
	});

	exports.log = function (ip, timestamp, playerCount) {
		var insertStatement = db.prepare('INSERT INTO pings (timestamp, ip, playerCount) VALUES (?, ?, ?)');

		db.serialize(function () {
			insertStatement.run(timestamp, ip, playerCount);
		});

		insertStatement.finalize();
	};

	exports.getTotalRecord = function (ip, callback) {
		var currentTime = util.getCurrentTimeMs();
		db.all("SELECT MAX(playerCount) FROM pings WHERE ip = ? AND timestamp >= ?", [ip, currentTime - config.RecordDuration], function (err, data) {
			callback(data[0]['MAX(playerCount)']);
		});
	};


	exports.getAverage = function (ip, callback) {
		var currentTime = util.getCurrentTimeMs();
		db.all("SELECT AVG(playerCount) FROM pings WHERE ip = ? AND timestamp >= ?", [ip, currentTime - config.AverageDuration], function (err, data) {
			//console.log(data)
			callback(data[0]['AVG(playerCount)']);
		});
	};

	exports.queryPings = function (duration, callback) {
		var currentTime = util.getCurrentTimeMs();

		db.all("SELECT * FROM pings WHERE timestamp >= ? AND timestamp <= ?", [
			currentTime - duration,
			currentTime
		], function (err, data) {
			callback(data);
		});
	};
};

assets/js/site.js
var graphs = [];
var lastPlayerEntries = [];

var historyPlot;
var displayedGraphData;
var hiddenGraphData = [];

var isConnected = false;

var mojangServicesUpdater;
var sortServersTask;

var currentServerHover;
var faviconSize = 64;

function updateServerStatus(lastEntry) {
    var info = lastEntry.info;

    var div = $('#status_' + safeName(info.name));
    var versionDiv = $('#version_' + safeName(info.name));

    if (lastEntry.versions) {
        var versions = '';

        for (var i = 0; i < lastEntry.versions.length; i++) {
            if (!lastEntry.versions[i]) continue;
            versions += '<span class="version">' + publicConfig.minecraftVersions[lastEntry.info.type][lastEntry.versions[i]] + '</span>&nbsp;';
        }

        versionDiv.html(versions);
    } else {
        versionDiv.html('');
    }

    if (lastEntry.result) {
        var result = lastEntry.result;
        var newStatus = 'Players: <span style="font-weight: 500;">' + formatNumber(result.players.online) + '</span>';

        var listing = graphs[lastEntry.info.name].listing;

        if (listing.length > 0) {
            newStatus += '<span class="color-gray"> (';

            var playerDifference = listing[listing.length - 1][1] - listing[0][1];

            if (playerDifference >= 0) {
                newStatus += '+';
            }

            newStatus += playerDifference + ')</span>';
        }

        lastPlayerEntries[info.name] = result.players.online;

        div.html(newStatus);
    } else {
        var newStatus = '<span class="color-red">';

        if (findErrorMessage(lastEntry.error)) {
            newStatus += findErrorMessage(lastEntry.error);
        } else {
            newStatus += 'Failed to ping!';
        }

        div.html(newStatus + '</span>');
    }

    var keys = Object.keys(lastPlayerEntries);
    var totalPlayers = 0;

    for (var i = 0; i < keys.length; i++) {
        totalPlayers += lastPlayerEntries[keys[i]];
    }

    $("#stat_totalPlayers").text(formatNumber(totalPlayers));
    $("#stat_networks").text(formatNumber(keys.length));

    if (lastEntry.record) {
        $('#record_' + safeName(info.name)).html('Record: ' + formatNumber(lastEntry.record));
        $('#average_' + safeName(info.name)).html('Average: ' + formatNumber(lastEntry.average.toFixed(2)) );
        console.log("-------------------------")
        console.log(info.name)
        console.log(lastEntry)
        console.log("-------------------------" + "\n\n\n\n")
        //console.log("Server: " + info.name + " Average: " + lastEntry.record)
    }

    updatePercentageBar();
}

function sortServers() {
    if (categoriesVisible) {
        var byCategories = getServersByCategory();

        var categories = Object.keys(byCategories);

        for (var i = 0; i < categories.length; i++) {
            var relevantPlayers = [];

            for (var x = 0; x < byCategories[categories[i]].length; x++) {
                var server = byCategories[categories[i]][x];

                relevantPlayers[server.name] = lastPlayerEntries[server.name];
            }

            var keys = Object.keys(relevantPlayers);

            keys.sort(function(a, b) {
                return relevantPlayers[b] - relevantPlayers[a];
            });

            for (var x = 0; x < keys.length; x++) {
                $('#container_' + safeName(keys[x])).appendTo('#server-container-' + categories[i]);
                $('#ranking_' + safeName(keys[x])).text('#' + (x + 1));
            }
        }
    } else {
        var serverNames = [];

        var keys = Object.keys(lastPlayerEntries);

        for (var i = 0; i < keys.length; i++) {
            serverNames.push(keys[i]);
        }

        serverNames.sort(function(a, b) {
            return (lastPlayerEntries[b] || 0) - (lastPlayerEntries[a] || 0);
        });

        for (var i = 0; i < serverNames.length; i++) {
            $('#container_' + safeName(serverNames[i])).appendTo('#server-container-all');
            $('#ranking_' + safeName(serverNames[i])).text('#' + (i + 1));
        }
    }
}

function updatePercentageBar() {
    var keys = Object.keys(lastPlayerEntries);

    keys.sort(function(a, b) {
        return lastPlayerEntries[a] - lastPlayerEntries[b];
    });

    var totalPlayers = getCurrentTotalPlayers();

    var parent = $('#perc-bar');
    var leftPadding = 0;

    for (var i = 0; i < keys.length; i++) {
        (function(pos, server, length) {
            var safeNameCopy = safeName(server);
            var playerCount = lastPlayerEntries[server];

            var div = $('#perc_bar_part_' + safeNameCopy);

            // Setup the base
            if (!div.length) {
                $('<div/>', {
                    id: 'perc_bar_part_' + safeNameCopy,
                    class: 'perc-bar-part',
                    html: '',
                    style: 'background: ' + getServerColor(server) + ';'
                }).appendTo(parent);

                div = $('#perc_bar_part_' + safeNameCopy);

                div.mouseover(function(e) {
                    currentServerHover = server;
                });

                div.mouseout(function(e) {
                    hideTooltip();
                    currentServerHover = undefined;
                });
            }

            // Update our position/width
            var width = (playerCount / totalPlayers) * parent.width();

            div.css({
                width: width + 'px',
                left: leftPadding + 'px'
            });

            leftPadding += width;
        })(i, keys[i], keys.length);
    }
}

function getCurrentTotalPlayers() {
    var totalPlayers = 0;
    var keys = Object.keys(lastPlayerEntries);
    for (var i = 0; i < keys.length; i++) totalPlayers += lastPlayerEntries[keys[i]]
    return totalPlayers;
}

function setAllGraphVisibility(visible) {
    if (visible) {
        var keys = Object.keys(hiddenGraphData);

        for (var i = 0; i < keys.length; i++) {
            displayedGraphData[keys[i]] = hiddenGraphData[keys[i]];
        }

        hiddenGraphData = [];
    } else {
        var keys = Object.keys(displayedGraphData);

        for (var i = 0; i < keys.length; i++) {
            hiddenGraphData[keys[i]] = displayedGraphData[keys[i]];
        }

        displayedGraphData = [];
    }

    $('.graph-control').each(function(index, item) {
        item.checked = visible;
    });

    historyPlot.setData(convertGraphData(displayedGraphData));
    historyPlot.setupGrid();

    historyPlot.draw();

    // Update our localStorage
    if (visible) {
        resetGraphControls();
    } else {
        saveGraphControls(Object.keys(displayedGraphData));
    }
}

function validateBootTime(bootTime, socket) {
    $('#tagline-text').text('Validating...');

    console.log('Remote bootTime is ' + bootTime + ', local is ' + publicConfig.bootTime);

    if (bootTime === publicConfig.bootTime) {
        $('#tagline-text').text('Loading...');

        socket.emit('requestListing');

        if (!isMobileBrowser()) socket.emit('requestHistoryGraph');

        isConnected = true;

        // Start any special updating tasks.
        mojangServicesUpdater = setInterval(updateMojangServices, 1000);
        sortServersTask = setInterval(sortServers, 10000);
    } else {
        $('#tagline-text').text('Updating...');

        $.getScript('/publicConfig.json', function(data, textStatus, xhr) {
            if (xhr.status === 200) {
                validateBootTime(publicConfig.bootTime, socket);
            } else {
                showCaption('Failed to update! Refresh?');
            }
        });
    }
}

function printPort(port) {
  if(port == undefined || port == 25565) {
    return "";
  } else {
    return ":" + port;
  }
}

$(document).ready(function() {
	var socket = io.connect({
        reconnect: true,
        reconnectDelay: 1000,
        reconnectionAttempts: 10
    });

    socket.on('bootTime', function(bootTime) {
        validateBootTime(bootTime, socket);
    });

    socket.on('disconnect', function() {
        if (mojangServicesUpdater) clearInterval(mojangServicesUpdater);
        if (sortServersTask) clearInterval(sortServersTask);

        lastMojangServiceUpdate = undefined;

        showCaption('Disconnected! Refresh?');

        lastPlayerEntries = {};
        graphs = {};

        $('#server-container-list').html('');

        createdCategories = false;

        $('#big-graph').html('');
        $('#big-graph-checkboxes').html('');
        $('#big-graph-controls').css('display', 'none');

        $('#perc-bar').html('');
        $('.mojang-status').css('background', 'transparent');
        $('.mojang-status-text').text('...');

        $("#stat_totalPlayers").text(0);
        $("#stat_networks").text(0);

        isConnected = false;
    });

    socket.on('historyGraph', function(rawData) {
        var shownServers = loadGraphControls();

        if (shownServers) {
            var keys = Object.keys(rawData);

            hiddenGraphData = [];
            displayedGraphData = [];

            for (var i = 0; i < keys.length; i++) {
                var name = keys[i];

                if (shownServers.indexOf(name) !== -1) {
                    displayedGraphData[name] = rawData[name];
                } else {
                    hiddenGraphData[name] = rawData[name];
                }
            }
        } else {
            displayedGraphData = rawData;
        }

        $('#big-graph').css('height', '400px');

        historyPlot = $.plot('#big-graph', convertGraphData(displayedGraphData), bigChartOptions);

        $('#big-graph').bind('plothover', handlePlotHover);

        var keys = Object.keys(rawData);

        var sinceBreak = 0;
        var html = '<table><tr>';

        keys.sort();

        for (var i = 0; i < keys.length; i++) {
            var checkedString = '';

            if (displayedGraphData[keys[i]]) {
                checkedString = 'checked=checked';
            }

            html += '<td><input type="checkbox" class="graph-control" id="graph-controls" data-target-network="' + keys[i] + '" ' + checkedString + '> ' + keys[i] + '</input></td>';

            if (sinceBreak >= 7) {
                sinceBreak = 0;

                html += '</tr><tr>';
            } else {
                sinceBreak++;
            }
        }

        $('#big-graph-checkboxes').append(html + '</tr></table>');
        $('#big-graph-controls').css('display', 'block');
    });

    socket.on('updateHistoryGraph', function(rawData) {
        // Prevent race conditions.
        if (!displayedGraphData || !hiddenGraphData) {
            return;
        }

        // If it's not in our display group, use the hidden group instead.
        var targetGraphData = displayedGraphData[rawData.name] ? displayedGraphData : hiddenGraphData;

        trimOldPings(targetGraphData, publicConfig.graphDuration);

        targetGraphData[rawData.name].push([rawData.timestamp, rawData.players]);

        // Redraw if we need to.
        if (displayedGraphData[rawData.name]) {
            historyPlot.setData(convertGraphData(displayedGraphData));
            historyPlot.setupGrid();

            historyPlot.draw();
        }
    });

	socket.on('add', function(servers) {
        for (var i = 0; i < servers.length; i++) {
            var history = servers[i];
            var listing = [];

            for (var x = 0; x < history.length; x++) {
                var point = history[x];

                if (point.result) {
                    listing.push([point.timestamp, point.result.players.online]);
                } else if (point.error) {
                    listing.push([point.timestamp, 0]);
                }
            }

            var lastEntry = history[history.length - 1];
            var info = lastEntry.info;

            if (lastEntry.error) {
                lastPlayerEntries[info.name] = 0;
            } else if (lastEntry.result) {
                lastPlayerEntries[info.name] = lastEntry.result.players.online;
            }

            var typeString = publicConfig.serverTypesVisible ? '<span class="type">' + info.type + '</span>' : '';

            var safeNameCopy = safeName(info.name);

            $('<div/>', {
                id: 'container_' + safeNameCopy,
                class: 'server',
                'server-id': safeNameCopy,
                html: '<div id="server-' + safeNameCopy + '" class="column" style="width: 80px;">\
                            <img id="favicon_' + safeNameCopy + '" title="' + info.name + '\n' + '" height="' + faviconSize + '" width="' + faviconSize + '">\
                            <br />\
                            <p class="text-center-align rank" id="ranking_' + safeNameCopy + '"></p>\
                        </div>\
                        <div class="column" style="width: 220px;">\
                            <h3>' + info.name + '&nbsp;' + typeString + '</h3>\
                            <span id="status_' + safeNameCopy + '">Waiting</span>\
                            <div <span class="version"></span></div>\
                            <div id="average_' + safeNameCopy + '" class="color-dark-gray server-meta"></span></div>\
                            <span id="record_' + safeNameCopy + '" class="color-dark-gray server-meta"></span>\
                        </div>\
                        <div class="column" style="float: right;">\
                            <div class="chart" id="chart_' + safeNameCopy + '"></div>\
                        </div>'
            }).appendTo("#server-container-" + getServerByIp(info.ip).category);

            var favicon = MISSING_FAVICON_BASE64;

            if (lastEntry.result && lastEntry.result.favicon) {
                favicon = lastEntry.result.favicon;
            }

            $('#favicon_' + safeName(info.name)).attr('src', favicon);

            graphs[lastEntry.info.name] = {
                listing: listing,
                plot: $.plot('#chart_' + safeNameCopy, [listing], smallChartOptions)
            };

            updateServerStatus(lastEntry);

            $('#chart_' + safeNameCopy).bind('plothover', handlePlotHover);
        }

        sortServers();
        updatePercentageBar();
	});

	socket.on('update', function(update) {
        // Prevent weird race conditions.
        if (!graphs[update.info.name]) {
            return;
        }

        // We have a new favicon, update the old one.
        if (update.result && update.result.favicon) {
            $('#favicon_' + safeName(update.info.name)).attr('src', update.result.favicon);
        }

        var graph = graphs[update.info.name];

        updateServerStatus(update);

        if (update.result) {
            graph.listing.push([update.info.timestamp, update.result ? update.result.players.online : 0]);

            if (graph.listing.length > 72) {
                graph.listing.shift();
            }

            graph.plot.setData([graph.listing]);
            graph.plot.setupGrid();

            graph.plot.draw();
        }
	});

	socket.on('updateMojangServices', function(data) {
        if (isConnected) {
            updateMojangServices(data);
        }
    });

    socket.on('syncComplete', function(data) {
        hideCaption();

        $(document).on('click', '.server', function(e) {
            var serverId = $(this).attr('server-id');
        });
    });

    $(document).on('click', '.graph-control', function(e) {
        var serverIp = $(this).attr('data-target-network');
        var checked = $(this).attr('checked');

        // Restore it, or delete it - either works.
        if (!this.checked) {
            hiddenGraphData[serverIp] = displayedGraphData[serverIp];

            delete displayedGraphData[serverIp];
        } else {
            displayedGraphData[serverIp] = hiddenGraphData[serverIp];

            delete hiddenGraphData[serverIp];
        }

        // Redraw the graph
        historyPlot.setData(convertGraphData(displayedGraphData));
        historyPlot.setupGrid();

        historyPlot.draw();

        // Update our localStorage
        if (Object.keys(hiddenGraphData).length === 0) {
            resetGraphControls();
        } else {
            saveGraphControls(Object.keys(displayedGraphData));
        }
    });

    $(document).on('mousemove', function(e) {
        if (currentServerHover) {
            var totalPlayers = getCurrentTotalPlayers();
            var playerCount = lastPlayerEntries[currentServerHover];

            renderTooltip(e.pageX + 10, e.pageY + 10, '<strong>' + currentServerHover + '</strong>: ' + roundToPoint(playerCount / totalPlayers * 100, 10) + '% of ' + formatNumber(totalPlayers) + ' tracked players.<br />(' + formatNumber(playerCount) + ' online.)');
        }
    });

    $(window).on('resize', function() {
        updatePercentageBar();

        if (historyPlot) {
            historyPlot.resize();
            historyPlot.setupGrid();
            historyPlot.draw();
        }
    });
});
